libraries/rush-lib/src/logic/operations/ShardedPhaseOperationPlugin.ts

```diff
@@ -0,0 +1,162 @@
+import type {
+  ICreateOperationsContext,
+  IPhasedCommandPlugin,
+  PhasedCommandHooks
+} from '../../pluginFramework/PhasedCommandHooks';
+import { Operation } from './Operation';
+import type { IOperationSettings, RushProjectConfiguration } from '../../api/RushProjectConfiguration';
+import { ShellOperationRunner } from './ShellOperationRunner';
+import {
+  getCustomParameterValuesByPhase,
+  getDisplayName,
+  getScriptToRun
+} from './ShellOperationRunnerPlugin';
+import { NullOperationRunner } from './NullOperationRunner';
+import { OperationStatus } from './OperationStatus';
+
+export const PLUGIN_NAME: 'ShardedPhasedOperationPlugin' = 'ShardedPhasedOperationPlugin';
+
+/**
+ * Phased command that shards a phase into multiple operations.
+ */
+export class ShardedPhasedOperationPlugin implements IPhasedCommandPlugin {
+  public apply(hooks: PhasedCommandHooks): void {
+    hooks.createOperations.tap(PLUGIN_NAME, spliceShards);
+  }
+}
+
+function spliceShards(existingOperations: Set<Operation>, context: ICreateOperationsContext): Set<Operation> {
+  const { projectConfigurations, rushConfiguration } = context;
+
+  const shardableOperations: Set<Operation> = new Set<Operation>();
+
+  for (const operation of existingOperations) {
+    const { associatedPhase: phase, associatedProject: project } = operation;
+    const operationSettings: IOperationSettings | undefined = getOperationSettings(phase, project);
+    if (operationSettings?.sharding) {
+      shardableOperations.add(operation);
+    }
+  }
+  const getCustomParameterValuesForPhase: (phase: IPhase) => ReadonlyArray<string> =
+    getCustomParameterValuesByPhase();
+
+  for (const operation of shardableOperations) {
+    const { associatedPhase: phase, associatedProject: project } = operation;
+    const operationSettings: IOperationSettings | undefined = getOperationSettings(phase, project);
+    if (phase && project && operationSettings?.sharding && operationSettings.sharding.count > 1) {
+      existingOperations.delete(operation);
+
+      const collatorNode: Operation = new Operation({
+        phase,
+        project
+      });
+      existingOperations.add(collatorNode);
+      const collatorDisplayName: string = `${getDisplayName(phase, project)} - collate`;
+      const rawScript: string | undefined = project.packageJson.scripts?.[`${phase.name}:collate`];
+
+      if (rawScript) {
+        const collatorRunner: ShellOperationRunner = new ShellOperationRunner({
+          commandToRun: rawScript,
+          displayName: collatorDisplayName,
+          phase,
+          rushConfiguration,
+          rushProject: project,
+          operationSettings
+        });
+        collatorNode.runner = collatorRunner;
+      } else {
+        collatorNode.runner = new NullOperationRunner({
+          name: collatorDisplayName,
+          result: OperationStatus.NoOp,
+          silent: phase.missingScriptBehavior === 'silent'
```

Comment:
We might need to have separate `missingScriptBehavior` values for the phase and for the collator. If you have a required phase that doesn't require collation, for example.

Reply:
I'll adjust to a separate `collatorScript` key in the sharding config in `rush-project.json` that lets you specify this.

Reply:
Added in [67598bf](https://github.com/microsoft/rushstack/pull/4652/commits/67598bf13fa258cef6cb08d8a3b6958ef958ca90)